text/0002-serializable-types.md

```diff
@@ -0,0 +1,553 @@
+
+MLAPI used to offer serialization support that is much less performant (due to boxing, runtime type checking etc.) and (arguably) less convenient. When [Standard RPC API](https://github.com/Unity-Technologies/com.unity.multiplayer.rfcs/blob/master/text/0001-std-rpc-api.md) was introduced, it also did open a whole set of opportunities in this area. At the time of writing this RFC, this proposed design applies to the serialization of RPC parameters and is very likely to be used in other areas that use serialization such as network variables.
+
+We want to offer built-in serialization support for most commonly used types and also support custom serialization for user-defined types with ease.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Serialization
+
+Multiplayer framework has built-in serialization support for C# and Unity primitive types out-of-the-box, also with ability to further extend network serialization for user-defined types implementing `INetworkSerializable` interface.
+
+### C# Primitives
+
+`bool`, `char`, `sbyte`, `byte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `float`, `double`, `string` types will be serialized by built-in serialization code.
+
+```cs
+[ServerRpc]
+void FooServerRpc(int somenumber, string sometext) { /* ... */ }
+
+void Update()
+{
+    if (Input.GetKeyDown(KeyCode.P))
+    {
+        FooServerRpc(Time.frameCount, "hello, world");
+    }
+}
+```
+
+### Unity Primitives
+
+`Color`, `Color32`, `Vector2`, `Vector3`, `Vector4`, `Quaternion`, `Ray`, `Ray2D` types will be serialized by built-in serialization code.
+
+```cs
+[ClientRpc]
+void BarClientRpc(Color somecolor) { /* ... */ }
+
+void Update()
+{
+    if (Input.GetKeyDown(KeyCode.P))
+    {
+        BarClientRpc(Color.red);
+    }
+}
+```
+
+### Enum Types
+
+A user-defined enum type will be serialized by built-in serialization code (with underlying integer type).
+
+```cs
+enum SmallEnum : byte
+{
+    A,
+    B,
+    C
+}
+
+enum NormalEnum // default -> int
+{
+    X,
+    Y,
+    Z
+}
+
+[ServerRpc]
+void ConfigServerRpc(SmallEnum smallEnum, NormalEnum normalEnum) { /* ... */ }
+
+void Update()
+{
+    if (Input.GetKeyDown(KeyCode.P))
+    {
+        ConfigServerRpc(SmallEnum.A, NormalEnum.X);
+    }
+}
+```
+
+### Static Arrays
+
+Static arrays like `int[]` will be serialized by built-in serialization code if their underlying type is either one of serialization supported types (e.g. `Vector3`) or if they implement `INetworkSerializable` interface.
+
+```cs
+[ServerRpc]
+void HelloServerRpc(int[] scores, Color[] colors) { /* ... */ }
+
+[ClientRpc]
+void WorldClientRpc(MyComplexType[] values) { /* ... */ }
+```
+
+### NetworkObject & NetworkBehaviour
+
+`NetworkObject` and `NetworkBehaviour` instances will be serialized by built-in serialization using their respective network object ID (`NetworkObject.NetworkId`) and network behaviour ID (`NetworkBehaviour.GetBehaviourId()`).
```

Comment:
very good point and I think I agree we should either serialize them or just not offer them and leave it up to dev to send `NetworkId` over if needed. what this proposes it sub-par. this feature is not new, it's actually migrating from existing MLAPI and making improvements on top of that (UNET also offered similar functionality). I think it's fair to say that we should leave this for later rather than putting it in early without thinking too much about what it means and what it would communicate.

long story short: I agree that we should leave this out for now.

Reply:
David pointed me to this 😃. I disagree on some of the points which have been made.
1. NetworkedObject isn't a class which can be serialized. It is just an identifier, it basically acts like an entity in an ECS.
2. This is a problem which can be solved. Solving this problem with magic numbers can work but I'd argue it is not good design for a framework. 

I totally agree though on the on the lifetime argument and that way Unity is solving this with it's Objects isn't great.

One way I think to get the best of the two worlds would be the following:

1. We have a NetworkedObjectId struct. That struct contains just a ulong field. When sending NetworkedObjects we send/receive that struct. As a user I can try to receive a NetworkedObject from this Id struct.  Benefits of this approach:
- When we designate a special number like 0 to "not found" we can expose this in a nice way with a "Found" property.
- The check whether a NetworkedObject exists will be done by the framework and not by the user. This means less boiler plate code and more opportunities to log issues. BitWriter/BitReader currently contains a bunch of very useful logging statements about missing NetworkedObjects.
- It's more extensible/maintainable, if we ever decide to change NetworkId from ulong to int we can do that without breaking user code, they won't even notice.

I'm still not entirely sold on this mainly because UNet/Mirror/PUN all use this pattern of serializing object because it is just really convenient and intuitive.

Reply:
@LukeStampfli — I actually do not have any strong opinions about this. However, I think it's out-of-scope for this RFC to standardize or propose a solution, that's why I'd like to leave it outside for now. (In fact, I implemented ID-based `NetworkObject` and `NetworkBehaviour` serialization over at MLAPI repo under `experimental/serializable-types` branch but removed it again). Let's have a separate discussion and RFC for this :)